File: prime-py/primeTextWay.py
from primeTextFactory import *
import prime
import math

folderPath = "prime-output"


def createPrimeTextFiles(num):
    dir = createAFolderForPrime(num)
    if dir == False:
        return f"Text files for primes till {num} already exist."

    data = prime.primeInRangeObj(1, num)
    for key in data:
        makeATextFileForPrime(num, key, data[key])


def primeInRangeObjTextWay(firstNum, lastNum, count=0):
    dir = createAFolderForPrime(lastNum)
    if dir == False:
        return f"Text files for primes till {lastNum} already exist."

    counter = count
    if count != 0:
        count = 0
    result = ''

    if firstNum < 3 and lastNum >= 3:
        result += "2, 3, "
        count += 2
    elif (firstNum >= 3 and firstNum < 5) and lastNum >= 3:
        result += "3, "
        count += 1

    checkIndex = math.ceil((firstNum-1) / 6)
    checkNum1 = 6 * checkIndex - 1
    checkNum2 = 6 * checkIndex + 1
    checkFlag = checkNum1 >= firstNum
    numberToInvestigate = checkNum1 if checkFlag else checkNum2
    numberToAdd = 2 if checkFlag else 4

    while numberToInvestigate <= lastNum:
        if prime.ifPrime(numberToInvestigate):
            result += f"{numberToInvestigate}, "
            count += 1
            if count == 10000:
                counter += 10000
                makeATextFileForPrime(lastNum, counter, result)
                count = 0
                result = ""

        numberToInvestigate += numberToAdd
        numberToAdd = 4 if numberToAdd == 2 else 2

    if count % 10000 != 0:
        makeATextFileForPrime(lastNum, counter + count, result)


def checkIfPrime(num):
    return isPrime(num)


def division(num):
    if num % 2 == 0:
        return 2
    if num % 3 == 0:
        return 3

    lastNumberToCheck = int(num**0.5)
    files = findAllFilesForDivision(lastNumberToCheck)

    for file in files:
        data = allNumbersInFileArr(file)
        for n in data:
            if num % int(n) == 0:
                return n

    return num


def allDivisions(num):
    result = {}

    while num != 1:
        div = division(num)
        result[div] = result[div] + 1 if result.get(div) else 1
        num //= int(div)

    return prime.scientificWayOfNum(result)


#  This function is working just when we have some prime text files
def primes(num):
    files = findAllFilesForDivision(num)

    if len(files):
        dir = createAFolderForPrime(num)

        if dir == False:
            return f"Text files for primes till {num} already exist."

        copyFiles(files, num, len(files)-1)

        data = allNumbersInFileArr(files[-1])
        res = [n for n in data if int(n) <= num]
        lastFile = files[-2] if len(files) > 1 else False
        lastCounter = int(lastFile.split("Output")[1]) if lastFile else False
        makeATextFileForPrime(num, lastCounter+len(res)
                              if lastCounter else len(res), ", ".join(res) + ", ")
    else:
        folder = findLastFolder()
        files = getSortedAllPrimeFilesInAFolder(folder)

        lastFolderNumber = int(folder.split("-")[-1])
        lastFileNumber = int((files[-1].split("Output")[1]).split(".txt")[0])
        primeInRangeObjTextWay(lastFolderNumber, num, lastFileNumber)
        copyFiles(files, num, len(files))


def divisionFromText(num):
    if num == 1:
        return -1

    checkNumber = int(num**0.5)
    files = findAllFilesForDivision(checkNumber)

    for file in files:
        data = allNumbersInFileArr(file)
        for n in data:
            if int(n) <= num and num % int(n) == 0:
                return n

    return num


def isPrimeFromText(num):
    return divisionFromText(num) == num


def allPrimesFromText(firstNum, lastNum, count=0):
    dir = createAFolderForPrime(lastNum)
    if dir == False:
        return f"Text files for primes till {lastNum} already exist."

    counter = count
    if not count:
        count = 0
    result = ''
    if firstNum < 3 and lastNum >= 3:
        result += "2, 3, "
        count += 2
    elif (firstNum >= 3 and firstNum < 5) and lastNum >= 3:
        result += "3, "
        count += 1

    checkIndex = math.ceil((firstNum-1) / 6)
    checkNum1 = 6 * checkIndex - 1
    checkNum2 = 6 * checkIndex + 1
    checkFlag = checkNum1 >= firstNum
    numberToInvestigate = checkNum1 if checkFlag else checkNum2
    numberToAdd = 2 if checkFlag else 4

    while numberToInvestigate <= lastNum:
        file = findTheProperFileForPrime(numberToInvestigate)
        if file:
            if checkIfPrime(numberToInvestigate):
                result += f"{numberToInvestigate}, "
                count += 1
        elif isPrimeFromText(numberToInvestigate):
            result += f"{numberToInvestigate}, "
            count += 1
            if count == 10000:
                counter += 10000
                makeATextFileForPrime(lastNum, counter, result)
                count = 0
                result = ''

        numberToInvestigate += numberToAdd
        numberToAdd = 4 if numberToAdd == 2 else 2

    if count % 10000 != 0:
        makeATextFileForPrime(lastNum, counter + count, result)


def copyPrimeFromExistedFiles(num, files):
    dir = createAFolderForPrime(num)
    if dir == False:
        return f"Text files for primes till {num} already exist."

    copyFiles(files, num, len(files)-1)
    data = allNumbersInFileArr(files[-1])
    res = [n for n in data if int(n) <= num]
    lastFile = files[-2] if len(files) > 1 else False
    lastCounter = int(lastFile.split("Output")[1]) if lastFile else False
    makeATextFileForPrime(num, lastCounter+len(res)
                          if lastCounter else len(res), ", ".join(res) + ", ")


def copyAndBuildPrimes(num, files, folder, lastFolderNumber, lastDivisionToCheck, counter):
    dir = createAFolderForPrime(num)
    if dir == False:
        return f"Text files for primes till {num} already exist."

    copyFiles(files, num, len(files))
    number = int(lastFolderNumber) + 1
    count = 0
    result = ""
    checkIndex = math.ceil((number-1)/6)
    checkNum1 = 6 * checkIndex - 1
    checkFlag = checkNum1 >= number
    numberToAdd = 2 if checkFlag else 4
    fs = findAllFilesForDivision(lastDivisionToCheck)

    while number <= num:
        flag = True
        for file in fs:
            data = allNumbersInFileArr(file)
            if int(data[-1]) > lastDivisionToCheck:
                data = [n for n in data if int(n) <= lastDivisionToCheck]

            for n in data:
                if number % int(n) == 0:
                    flag = False
                    break

            if not flag:
                break

        if flag:
            result += f"{number}, "
            count += 1
            if count == 10000:
                counter += 10000
                makeATextFileForPrime(num, counter, result)
                count = 0
                result = ""

        number += numberToAdd
        numberToAdd = 4 if numberToAdd == 2 else 2

    if count % 10000 != 0:
        makeATextFileForPrime(num, counter + count, result)


def buildPrimes(num, numberToCreateFiles, files, folder, lastFolderNumber, counter):
    while numberToCreateFiles <= num:
        dir = createAFolderForPrime(numberToCreateFiles)
        if dir == False:
            return f"Text files for primes till {numberToCreateFiles} already exist."

        copyFiles(files, numberToCreateFiles, len(files))
        number = int(lastFolderNumber) + 1
        count = 0
        result = ""
        checkIndex = math.ceil((number - 1) / 6)
        checkNum1 = 6 * checkIndex - 1
        checkFlag = checkNum1 >= number
        numberToAdd = 2 if checkFlag else 4

        while number <= numberToCreateFiles:
            flag = True
            lastDivisionToCheck = int(number ** 0.5)
            fs = findAllFilesForDivision(lastDivisionToCheck)

            for file in fs:
                data = allNumbersInFileArr(file)
                if int(data[-1]) > number:
                    data = [n for n in data if int(n) <= number]

                for n in data:
                    if number % int(n) == 0:
                        flag = False
                        break

                if not flag:
                    break

            if flag:
                result += f"{number}, "
                count += 1
                if count == 10000:
                    counter += 10000
                    makeATextFileForPrime(numberToCreateFiles, counter, result)
                    count = 0
                    result = ""

            number += numberToAdd
            numberToAdd = 4 if numberToAdd == 2 else 2

        if count % 10000 != 0:
            makeATextFileForPrime(numberToCreateFiles, counter+count, result)

        folder = findLastFolder()
        files = getSortedAllPrimeFilesInAFolder(folder)
        lastFolderNumber = int(folder.split("-")[-1])
        counter = int((files[-1].split("Output")[1]).split(".")[0])
        numberToCreateFiles = lastFolderNumber ** 2

    if numberToCreateFiles > num:
        allPrimes(num)


def allPrimes(num):
    files = findAllFilesForDivision(num)
    if len(files) > 0:
        copyPrimeFromExistedFiles(num, files)
    else:
        lastDivisionToCheck = int(num ** 0.5)
        folder = findLastFolder()
        files = getSortedAllPrimeFilesInAFolder(folder)
        lastFolderNumber = int(folder.split('-')[-1])
        counter = int((files[-1].split('Output')[1]).split(".")[0])

        if lastDivisionToCheck <= lastFolderNumber:
            copyAndBuildPrimes(num, files, folder,
                               lastFolderNumber, lastDivisionToCheck, counter)
        else:
            numberToCreateFiles = lastFolderNumber ** 2
            buildPrimes(num, numberToCreateFiles, files,
                        folder, lastFolderNumber, counter)


def allPrimesSecondWay(num):
    files = findAllFilesForDivision(num)
    if len(files) > 0:
        copyPrimeFromExistedFiles(num, files)
    else:
        folder = findLastFolder()
        files = getSortedAllPrimeFilesInAFolder(folder)
        lastFolderNumber = int(folder.split("-")[-1])
        counter = int((files[-1].split("Output")[1]).split(".")[0])
        primeInRangeObjTextWay(lastFolderNumber + 1, num, counter)
        copyFiles(files, num, len(files))

# allPrimesFromText(1, 200000)
# print(int(65465564729**0.5))
# print(isPrimeFromText(65465564729))
# print(divisionFromText(77))
# primes(10)
# print(allDivisions(6547117 * 65465566399))
# print(division(337))
# print(checkIfPrime(25))
# primeInRangeObjTextWay(1, 2000000)
# print(createPrimeTextFiles(200))
# print(allPrimes(1000000000))

File: prime-js/primeTextWay.js
const {
  findTheProperFileForPrime,
  makeATextFileForPrime,
  createAFolderForPrime,
  isPrime,
  allNumbersInFileArr,
  findAllFilesForDivision,
  findLastFolder,
  copyFiles,
  sortFilesForPrimeInFolder,
} = require("./primeTextFactory");

const {
  ifPrime,
  primeInRangeObj,
  scientificWayOfNum,
} = require("./primeFirstWay");

const folderPath = "prime-output";

const createPrimeTextFiles = (num) => {
  let dir = createAFolderForPrime(num);
  if (!dir) {
    return `Text files for primes till ${num} already exist.`;
  }
  let data = primeInRangeObj(1, num);
  for (let key in data) {
    makeATextFileForPrime(num, key, data[key]);
  }
};

const primeInRangeObjTextWay = (firstNum, lastNum, count = 0) => {
  let dir = createAFolderForPrime(lastNum);
  if (!dir) {
    return `Text files for primes till ${lastNum} already exist.`;
  }

  let counter = count;
  if (count !== 0) count = 0;
  let result = "";
  if (firstNum < 3 && lastNum >= 3) {
    result += "2, 3, ";
    count += 2;
  } else if ((firstNum >= 3 && firstNum < 5) & (lastNum >= 3)) {
    result += "3, ";
    count += 1;
  }

  let checkIndex = Math.ceil((firstNum - 1) / 6);
  let checkNum1 = 6 * checkIndex - 1;
  let checkNum2 = 6 * checkIndex + 1;
  let checkFlag = checkNum1 >= firstNum;
  let numberToInvestigate = checkFlag ? checkNum1 : checkNum2;
  let numberToAdd = checkFlag ? 2 : 4;

  while (numberToInvestigate <= lastNum) {
    if (ifPrime(numberToInvestigate)) {
      result += `${numberToInvestigate}, `;
      count++;
      if (count === 10000) {
        counter += 10000;
        makeATextFileForPrime(lastNum, counter, result);
        count = 0;
        result = "";
      }
    }

    numberToInvestigate += numberToAdd;
    numberToAdd = numberToAdd === 2 ? 4 : 2;
  }
  if (count % 10000 !== 0)
    makeATextFileForPrime(lastNum, counter + count, result);
};

const createPrimeTextFilesRange = (firstNum, lastNum) => {
  const folderPath = "prime-between-output";
  const fileName = firstNum + "-" + lastNum;
  let dir = createAFolderForPrime(fileName, folderPath);

  if (!dir) {
    return `Text files for primes till ${fileName} already exist.`;
  }

  let data = primeInRangeObj(firstNum, lastNum);
  for (let key in data) {
    makeATextFileForPrime(fileName, key, data[key], folderPath);
  }
};

const checkIfPrime = (num) => {
  return isPrime(num);
};

const division = (num) => {
  if (!(num % 2)) return 2;
  if (!(num % 3)) return 3;

  const lastNumberToCheck = Math.floor(Math.sqrt(num));
  let files = findAllFilesForDivision(lastNumberToCheck);

  for (let f of files) {
    let data = allNumbersInFileArr(f);
    for (let n of data) {
      if (num % n === 0) {
        return n;
      }
    }
  }
  return num;
};

const allDivisions = (num) => {
  let result = {};
  while (num != 1) {
    let div = division(num);
    result[div] ? result[div]++ : (result[div] = 1);
    num /= div;
  }
  return scientificWayOfNum(result);
};

// This function is working just when we have some prime text files.
const primes = (num) => {
  let files = findAllFilesForDivision(num);

  if (files.length) {
    let dir = createAFolderForPrime(num);
    if (!dir) {
      return `Text files for primes till ${num} already exist.`;
    }
    copyFiles(files, num, files, files.length - 1);
    let data = allNumbersInFileArr(files[files.length - 1]);
    let res = data.filter((n) => n <= num);
    let lastFile = files[files.length - 2];
    let lastCounter = lastFile
      ? parseInt(files[files.length - 2].split("Output")[1])
      : false;
    makeATextFileForPrime(
      num,
      lastCounter ? lastCounter + res.length : res.length,
      res.join(", ") + ", "
    );
  } else {
    let folder = findLastFolder();
    files = sortFilesForPrimeInFolder(folder);
    let lastFolderNumber = folder.split("-")[1];
    let lastFileNumber = parseInt(files[files.length - 1].split("Output")[1]);
    primeInRangeObjTextWay(+lastFolderNumber, num, +lastFileNumber);
    copyFiles(files, num, folder, files.length);
  }
};

const divisionFromText = (num) => {
  if (num === 1) return -1;
  let checkNumber = Math.floor(Math.sqrt(num));
  let files = findAllFilesForDivision(checkNumber);
  for (let f of files) {
    let data = allNumbersInFileArr(f);
    for (let n of data) {
      if (n <= num && num % n === 0) {
        return n;
      }
    }
  }
  return num;
};

const isPrimeFromText = (num) => {
  return divisionFromText(num) == num;
};

const allPrimesFromText = (firstNum, lastNum, count = 0) => {
  let dir = createAFolderForPrime(lastNum);
  if (!dir) {
    return `Text files for primes till ${lastNum} already exist.`;
  }

  let counter = count;
  if (!count) count = 0;
  let result = "";
  if (firstNum < 3 && lastNum >= 3) {
    result += "2, 3, ";
    count += 2;
  } else if ((firstNum >= 3 && firstNum < 5) & (lastNum >= 3)) {
    result += "3, ";
    count += 1;
  }

  let checkIndex = Math.ceil((firstNum - 1) / 6);
  let checkNum1 = 6 * checkIndex - 1;
  let checkNum2 = 6 * checkIndex + 1;
  let checkFlag = checkNum1 >= firstNum;
  let numberToInvestigate = checkFlag ? checkNum1 : checkNum2;
  let numberToAdd = checkFlag ? 2 : 4;
  while (numberToInvestigate <= lastNum) {
    let file = findTheProperFileForPrime(numberToInvestigate);
    if (file) {
      if (checkIfPrime(numberToInvestigate)) {
        result += `${numberToInvestigate}, `;
        count++;
      }
    } else if (isPrimeFromText(numberToInvestigate)) {
      result += `${numberToInvestigate}, `;
      count++;
      if (count === 10000) {
        counter += 10000;
        makeATextFileForPrime(lastNum, counter, result);
        count = 0;
        result = "";
      }
    }

    numberToInvestigate += numberToAdd;
    numberToAdd = numberToAdd === 2 ? 4 : 2;
  }
  if (count % 10000 !== 0)
    makeATextFileForPrime(lastNum, counter + count, result);
};

// this method used in `allPrimes` function
// if a number that we want to find all the primes till that is less -
// - than primes on existed folders, this will be run
const copyPrimeFromExistedFiles = (num, files) => {
  let dir = createAFolderForPrime(num);
  if (!dir) {
    return `Text files for primes till ${num} already exist.`;
  }
  copyFiles(files, num, files, files.length - 1);
  let data = allNumbersInFileArr(files[files.length - 1]);
  let res = data.filter((n) => n <= num);
  let lastFile = files[files.length - 2];
  let lastCounter = lastFile
    ? parseInt(files[files.length - 2].split("Output")[1])
    : false;
  makeATextFileForPrime(
    num,
    lastCounter ? lastCounter + res.length : res.length,
    res.join(", ") + ", "
  );
};

// this method used in `allPrimes` function
// if a number that we want to find all the primes till that is bigger -
// - than primes on existed folders, but sqrt of that is less than existed primes, -
// - this will copy all the existed ones and will find the primes from greatest existed prime -
// till the number.
const copyAndBuildPrimes = (
  num,
  files,
  folder,
  lastFolderNumber,
  lastDivisionToCheck,
  counter
) => {
  let dir = createAFolderForPrime(num);
  if (!dir) {
    return `Text files for primes till ${num} already exist.`;
  }
  copyFiles(files, num, folder, files.length);
  let number = +lastFolderNumber + 1;
  let count = 0;
  let result = "";
  let checkIndex = Math.ceil((number - 1) / 6);
  let checkNum1 = 6 * checkIndex - 1;
  let checkFlag = checkNum1 >= number;
  let numberToAdd = checkFlag ? 2 : 4;
  let fs = findAllFilesForDivision(lastDivisionToCheck);
  while (number <= num) {
    let flag = true;
    for (let file of fs) {
      let data = allNumbersInFileArr(file);
      if (data[data.length - 1] > lastDivisionToCheck) {
        data = data.filter((n) => n <= lastDivisionToCheck);
      }
      for (let n of data) {
        if (number % n === 0) {
          flag = false;
          break;
        }
      }

      if (!flag) break;
    }
    if (flag) {
      result += `${number}, `;
      count++;
      if (count === 10000) {
        counter += 10000;
        makeATextFileForPrime(num, counter, result);
        count = 0;
        result = "";
      }
    }

    number += numberToAdd;
    numberToAdd = numberToAdd === 2 ? 4 : 2;
  }
  if (count % 10000 !== 0) makeATextFileForPrime(num, counter + count, result);
};

// this method used in `allPrimes` function
// if a number that we want to find all the primes till that is bigger -
// - than primes on existed folders, and sqrt of that is bigger than existed primes, -
// - this will copy all the existed ones and will build all the primes till (existed maximum prime)**2, -
// - and then will repeat this process till the folders to build the primes for number are existed.
const buildPrimes = (
  num,
  numberToCreateFiles,
  files,
  folder,
  lastFolderNumber,
  counter
) => {
  while (numberToCreateFiles <= num) {
    let dir = createAFolderForPrime(numberToCreateFiles);
    if (!dir) {
      return `Text files for primes till ${numberToCreateFiles} already exist.`;
    }
    copyFiles(files, numberToCreateFiles, folder, files.length);
    let number = +lastFolderNumber + 1;
    let count = 0;
    let result = "";
    let checkIndex = Math.ceil((number - 1) / 6);
    let checkNum1 = 6 * checkIndex - 1;
    let checkFlag = checkNum1 >= number;
    let numberToAdd = checkFlag ? 2 : 4;

    while (number <= numberToCreateFiles) {
      let flag = true;
      let lastDivisionToCheck = Math.floor(Math.sqrt(number));
      let fs = findAllFilesForDivision(lastDivisionToCheck);

      for (let file of fs) {
        let data = allNumbersInFileArr(file);
        if (data[data.length - 1] > number) {
          data = data.filter((n) => n <= number);
        }
        for (let n of data) {
          if (number % n === 0) {
            flag = false;
            break;
          }
        }

        if (!flag) break;
      }
      if (flag) {
        result += `${number}, `;
        count++;
        if (count === 10000) {
          counter += 10000;
          makeATextFileForPrime(numberToCreateFiles, counter, result);
          count = 0;
          result = "";
        }
      }

      number += numberToAdd;
      numberToAdd = numberToAdd === 2 ? 4 : 2;
    }
    if (count % 10000 !== 0)
      makeATextFileForPrime(numberToCreateFiles, counter + count, result);
    folder = findLastFolder();
    files = sortFilesForPrimeInFolder(folder);
    lastFolderNumber = folder.split("-")[1];
    counter = parseInt(files[files.length - 1].split("Output")[1]);

    numberToCreateFiles = lastFolderNumber ** 2;
  }
  if (numberToCreateFiles > num) {
    allPrimes(num);
  }
};

const allPrimes = (num) => {
  let files = findAllFilesForDivision(num);
  if (files.length) {
    copyPrimeFromExistedFiles(num, files);
  } else {
    let lastDivisionToCheck = Math.floor(Math.sqrt(num));

    let folder = findLastFolder();
    files = sortFilesForPrimeInFolder(folder);
    let lastFolderNumber = folder.split("-")[1];
    let counter = parseInt(files[files.length - 1].split("Output")[1]);

    if (lastDivisionToCheck <= lastFolderNumber) {
      copyAndBuildPrimes(
        num,
        files,
        folder,
        lastFolderNumber,
        lastDivisionToCheck,
        counter
      );
    } else {
      let numberToCreateFiles = lastFolderNumber ** 2;
      buildPrimes(
        num,
        numberToCreateFiles,
        files,
        folder,
        lastFolderNumber,
        counter
      );
    }
  }
};

const allPrimesSecondWay = (num) => {
  let files = findAllFilesForDivision(num);
  if (files.length) {
    copyPrimeFromExistedFiles(num, files);
  } else {
    let folder = findLastFolder();
    files = sortFilesForPrimeInFolder(folder);
    let lastFolderNumber = folder.split("-")[1];
    let counter = parseInt(files[files.length - 1].split("Output")[1]);
    primeInRangeObjTextWay(+lastFolderNumber + 1, num, counter);
    copyFiles(files, num, folder, files.length);
  }
};

module.exports = {
  primeInRangeObjTextWay,
  checkIfPrime,
  division,
  allDivisions,
  primes,
  divisionFromText,
  isPrimeFromText,
  allPrimesFromText,
  allPrimes,
  isPrime,
  ifPrime,
  primeInRangeObj,
  allPrimesSecondWay,
};
